### PR TITLE
reset password

### PR DESCRIPTION
**Pull request by SahithiKotte11**

### CODE REVIEW

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 1]
```diff
@@ -143,15 +152,34 @@ public LoginResponse refreshToken(final String refreshToken) {
...
   @Override
-  @Transactional(readOnly = true)
+  @Transactional(readOnly = false)
   public void sendResetPasswordEmail(String email) {
-    userService.findByEmail(email);
+    User user = userService.findByEmail(email);
+
+    if (user == null) {
+      throw new GlobalExceptionHandler.GenericException("User not found");
+    }
```
Comment (nnagwani-traitfit) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:161]: Why are we throwing a generic exception, when a better exception exists for this.

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 2]
```diff
@@ -143,15 +152,34 @@ public LoginResponse refreshToken(final String refreshToken) {
...
 
-    awsService.sendEmail(senderEmail, email, "cc_reset_password", Map.of("link", resetPasswordUrl));
+    Optional<ResetPassword> existingTokenOpt = resetPasswordRepo.findByUserIdAndLinkSent(user.getUserId());
+    if (existingTokenOpt.isPresent()) {
+      ResetPassword existingToken = existingTokenOpt.get();
+      if (Duration.between(existingToken.getCreatedAt(), LocalDateTime.now()).toHours() < 1) {
+        throw new GlobalExceptionHandler.GenericException("A password reset link has already been sent. Please try again later.");
+      }
```
Comment (nnagwani-traitfit) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:168]: Same. This is a specific problem, can we throw a specific exception.

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 3]
```diff
@@ -143,15 +152,34 @@ public LoginResponse refreshToken(final String refreshToken) {
...
+    String token = UUID.randomUUID().toString();
+    ResetPassword resetToken = new ResetPassword();
+    resetToken.setUserId(user.getUserId());
+    resetToken.setResetToken(token);
+    resetToken.setLinkSent(true);
+    resetPasswordRepo.save(resetToken);
+
+
+    awsService.sendEmail(senderEmail, email, "cc_reset_password", Map.of("link", resetPasswordUrl+"?token=" + token));
   }
```
Comment (nnagwani-traitfit) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:180]: What if this throws an exception. Line 180

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 4]
```diff
@@ -191,6 +219,29 @@ public void createNewUser(UserCreateRequest userCreateRequest) {
     passwordHistoryRepository.saveAndFlush(passwordHistory);
   }
 
+  @Override
+  public void resetPassword(ResetPasswordRequest resetPasswordRequest) {
+    Optional<ResetPassword> resetToken = resetPasswordRepo.findByResetToken(resetPasswordRequest.getResetToken());
+    if(resetToken.isPresent()){
+      if(resetToken.get().isLinkExpired()){
+        throw new GlobalExceptionHandler.GenericException("Link has expired. Please request a new link.");
+      }
```
Comment (nnagwani-traitfit) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:228]: Same problem as above

**Review by nnagwani-traitfit — COMMENTED** [review]
Comments on the review to be addressed. other than that looks good.

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 5]
```diff
@@ -191,6 +219,29 @@ public void createNewUser(UserCreateRequest userCreateRequest) {
     passwordHistoryRepository.saveAndFlush(passwordHistory);
   }
 
+  @Override
+  public void resetPassword(ResetPasswordRequest resetPasswordRequest) {
+    Optional<ResetPassword> resetToken = resetPasswordRepo.findByResetToken(resetPasswordRequest.getResetToken());
+    if(resetToken.isPresent()){
```
Comment (BhargavaNarendra97) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:225]: Can you make it a single condition like if((resetToken.present() &&resetToken.get().isLinkExpired()) || resetToken.isEmpty()) {
throw new GlobalExceptionHandler.GenericException("Link has expired. Please request a new link.");
}

**Review by BhargavaNarendra97 — CHANGES_REQUESTED** [review]

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SahithiKotte11) [reply src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:225]: An error is thrown saying Invalid Token

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java (hunk 3 above)
Comment (SahithiKotte11) [reply src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:180]: it is handled in awsService itself.

File: src/main/java/com/mpsp/cc_auth_service/repository/PasswordHistoryRepo.java [hunk 6]
```diff
@@ -4,11 +4,16 @@
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
```
Comment (BhargavaNarendra97) [inline src/main/java/com/mpsp/cc_auth_service/repository/PasswordHistoryRepo.java:7]: Remove unused imports

File: src/main/java/com/mpsp/cc_auth_service/utils/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -81,10 +82,12 @@ public ResponseEntity<ErrorResponse> handleGenericException(Exception ex) {
     return new ResponseEntity<>(errorResponse, HttpStatus.INTERNAL_SERVER_ERROR);
   }
 
+
+
   @ExceptionHandler(SesV2Exception.class)
```
Comment (BhargavaNarendra97) [inline src/main/java/com/mpsp/cc_auth_service/utils/GlobalExceptionHandler.java:87]: If we are not throwing SesV2Exception anywhere, is this handler still required?

File: src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java [hunk 8]
```diff
@@ -191,6 +218,26 @@ public void createNewUser(UserCreateRequest userCreateRequest) {
     passwordHistoryRepository.saveAndFlush(passwordHistory);
   }
 
+  @Override
+  public void resetPassword(ResetPasswordRequest resetPasswordRequest) {
+    Optional<ResetPassword> resetToken = resetPasswordRepo.findByResetToken(resetPasswordRequest.getResetToken());
```
Comment (BhargavaNarendra97) [inline src/main/java/com/mpsp/cc_auth_service/service/impl/AuthServiceImpl.java:223]: Please simplify it even further using the Optional orElseThrow method.

**Review by BhargavaNarendra97 — CHANGES_REQUESTED** [review]

**Pull request merged**